### ISSUE TITLE
Use `![alt](path)` syntax not only to include images, but also text files

### ISSUE THREAD

**Issue by AvtechScientific** [jgm/djot#307]
Use `![alt](path)` syntax not only to include images, but also text files. 

See: https://github.com/jgm/djot.js/issues/85

"This commit enables the creation of structured documents of arbitrary complexity, contrary to the current state of affairs where only one-pagers are allowed. So now if one needs to write a book or a lengthy article all the content must be in one file. Working with big files is not convenient and may slow down/crash the editor besides being hard to read. File inclusion, implemented in https://github.com/jgm/djot.js/pull/94, allows the author to subdivide the book/article into separate manageable chapters. This provides djot with the power similar to that of LaTeX and makes it positively distinct from all the Markdown-like tools."

And a ready to merge PR:  https://github.com/jgm/djot.js/pull/94

**Comment by Omikhleia** [comment]
Sorry, but there's a discrepancy between this description and what the code does. I see some code related to footnotes, etc. So what does it mean? What are the implications for other Djot implementations?

**Comment by Omikhleia** [comment]
Also, sorry, but this is bragging, and weird:
> This provides djot with the power similar to that of LaTeX and makes it positively distinct from all the Markdown-like tools.

In my own workflows, I use a higher level master document file (above Djot etc.), and I have made printed books from back cover to front cover with them, assembling bits of Djot content (and also Markdown or other, when needed). I don't see anything being solved positively here.

**Comment by AvtechScientific** [comment]
If you include files - there might be colliding identifiers, like e.g.
footnote names. In such a case the renderer warns the user and it is up to
the user to handle the issue. Again - no manipulation with the content.

**Comment by AvtechScientific** [comment]
> In my own workflows, I use a higher level master document file (above Djot etc.), and I have made printed books from front cover to back cover with them, assembling bits of Djot content (and also Markdown or other, when needed). I don't see anything being solved "positively" here.

Using your own workflows can you handle this simple case of two included files:

File 1:
Is this **bold

File 2:
or not?**... hmm

If yes - how do you do it?

**Comment by Omikhleia** [comment]
I certainly would _forbid_ breaking _semantic_ structures between different files. What your example above is even aiming at _solving_ ? You do really expect someone to start a bold structure in a file... and end it in another, seriously?

**Comment by Omikhleia** [comment]
Speaking of semantics, what becomes the "alt" content in your `![alt](somefile.djot)` ?

**Comment by AvtechScientific** [comment]
> Speaking of semantics, what becomes the "alt" content in your `![alt](somefile.djot)` ?

Just to stay consistent with the image case - `alt` will be displayed if `somefile.djot` is not found.

**Comment by AvtechScientific** [comment]
> I certainly would _forbid_ breaking _semantic_ structures between different files. What your example above is even aiming at _solving_ ? You do really expect someone to start a bold structure in a file... and end it in another, seriously?

1. Judging by the lack of a solution on your side I conclude that you have no solution.
2. The above mentioned example was intended for demonstration purposes only. If you can't think of a practical example for this - I'll help you. Inclusion, i.a., enables templates. Let's imagine a document that consists of 2 constant sections - header and footer, and a variable middle section (e.g. a letter to company employees were the variable middle section consists of employee names being programmatically injected). Header might open an inline element (bold, italic or whatever) that footer will close. It's just one example. I think the real life can bring more.

**Comment by Omikhleia** [comment]
> Just to stay consistent with the image case - alt will be displayed if somefile.djot is not found.

How consistent? In pandoc-flavored markdown, what you call the "alt" text may be used as the figure legend if the image stands alone in a paragraph of its own (see the "implicit figure" option).
The Djot syntax is not very clear _yet_ about this use case, and it does not have a general provision either for captioned images and figures -- for reference, see notably discussions https://github.com/jgm/djot/issues/28 and https://github.com/jgm/djot/issues/28, amongst other. But whether it eventually goes the same way as in pandoc (using the bracketed text as implicit figure) or via a generalization of the (currently table-only) `^` legend markup, the issue would remain the same.

> Judging by the lack of a solution on your side I conclude that you have no solution.

Sympathetic. But I recognize I wouldn't have a solution for a non-issue edge case with no clear semantics defined ;)

You mention templating - I'm not even that sure it should be part of the document syntax... And real templating goes far beyond mere content injection.
(For the mere record, however, I already need and use some sort of custom templating logic too, see https://github.com/jgm/djot/issues/238, in specialized Djot-based template files. I do think the real life can bring more, indeed.)

**Comment by criloz** [comment]
I think the association of this syntax `![alt](path)`  with images is so widespread that it does not worth the effort to make it different.

An option could just use other symbol like `#`

1. `#(table.csv)`, by default use the file extension to detect the file type
2. `#[csv](table.csv)`, file type can be overwritten like this  and ignore file extension
3. `#[=](table.csv)` |`#[=csv](table.csv)`, indicate in the AST that the file should be included as raw content, the file type could be used to potentially add some kind of syntax highlight
4.  `#(table.csv@12..34)`  include only a slice  of the file
5.  `#(profile.djot#who-i-am)`  include a section of the file that can be addressed by references

the only issue, would be that this symbol `#`, is used for tagging in certain system, but generally tags don't use symbols like `[`, `]`, `(`, `)`, or need to start by a valid `unicode_id_start` character.

**Comment by Omikhleia** [comment]
Your (4) goes far beyond the Djot input format markup and is perhaps possibly best left to renderers' interpretation... If instead of a CSV, your'd have (to stay on something similar) a speadsheet (ODT/OOXML), the "Sheet" name might be needed, etc. All of this is highly dependent on the source format, so a specification would have to be very explicit. (What's a slice of a CSV actually? Lines, columns, both? Etc.)

**Comment by criloz** [comment]
@Omikhleia  you are right on this and it was mostly related with other comment related to this https://github.com/jgm/djot/issues/199#issuecomment-1660893809, but this can also solve using URIs  `#(table.csv?start_row=12&end_row=34)` which let more freedom to the interpreter how to handle this

**Comment by vassudanagunta** [comment]
> I think the association of this syntax ![alt](path) with images is so widespread that it does not worth the effort to make it different.

It would not be making it different, but would be generalizing it, with consistent [transclusion](https://en.m.wikipedia.org/wiki/Transclusion) semantics for any referenced media type:

- `[label](moon.jpg)`: link to image
- `![alt](moon.jpg)`: embed image
- `[label](phases.csv)`: link to CSV
- `![alt](phases.csv)`: embed CSV as table

**Comment by criloz** [comment]
@vassudanagunta I think that the problem with that is what would be the meaning of `[alt]`?, in images and links is used as label, but in a csv or djot file, etc? Also, how do you assign a file-type, just using the file extension is not enough because there are some file extension that clashes, and in other cases the file extension is not even present at all like an URL `https://example.com/user/81a6bf136427d9e/raw/acde69adea3db6`

other option will allow using attributes to sort this out

- `![](https://example.com/user/81a6bf136427d9e/raw/acde69adea3db6){file_type=csv}`

**Comment by Omikhleia** [comment]
>     * `![](https://example.com/user/81a6bf136427d9e/raw/acde69adea3db6){file_type=csv}`

Another way perhaps, which remains compatible with the default/current syntax: `!format[alt](url)` where "format" is optional and "guessed" by the rendered (btw. the file extension is not the only way, on could do a file introspection). This would allow for `!csv[alt](myfile.txt)`

**Comment by criloz** [comment]
> > ```
> > * `![](https://example.com/user/81a6bf136427d9e/raw/acde69adea3db6){file_type=csv}`
> > ```
> 
> Another way perhaps, which remains compatible with the default/current syntax: `!format[alt](url)` where "format" is optional and "guessed" by the renderer (btw. the file extension is not the only way, one could do a file introspection, etc.). This would allow for `!csv[alt](myfile.txt)`

This `!format[alt](uri)` actually could work, and be implemented easily at the tokenizer level,   where the format could be a valid Unicode identifier, it will match `!{unicode_id}[` as token to start the file transclusion otherwise will just emit a text token.

But there are 2 types of files, transclusion

1. include the file and render it (interpreted), example a csv file end it as table
2. include the file content without interpretation, it could be the equivalent to write a code block with content of the file.

So based on that and to maintain consistency, `djot` can potentially handle both cases and use a similar syntax to what is used currently to differentiate between a raw content and a code block

1. `!=[alt](uri)`, interpreted without a format
2. `![alt](uri)`, include the file content without a format
3.  `!=format[alt](uri)`, interpreted with a format
4.  `!format[alt](uri)`, include the file content with a format, allowing syntax highlight

the problem is that this will change the current semantic of `![alt](uri)`, but an exception could be made to Uris that have an image file type to maintain some kind of backward compatibility

**Comment by criloz** [comment]
@Omikhleia you are right, the djot class system can be used for this, there are many ways to interpret a file

**Comment by mklcp** [comment]
I second using the class system. Even the distinction between `![alt](uri)` and `!=[alt](uri)` can be handled by the class system. And it is not always meaningful, e.g. `!=[alt](uri)`. What does that mean? Literally include the image as a stream of bytes? Written in hex? in rgb? Continuous or interspersed with spaces between columns (like hexdump)?

I'd rather have `![alt](uri)` generalized to all kind of files, and use classes to customize if needed or if ambiguous.

**Comment by criloz** [comment]
@mklcp I agree about the class system, even better because attributes do accept key-vaue pairs, it can potentially mimic function invocation, which allow great versatility and extensibility 

but I think it will be better if there is an unambiguous way to declare the file type, so the render engine will have 3 variables to dispatch the proper method

1. intention to include the file ![alt](path)
2. the file type, in cases where is impossible to determine the file type from the uri/path, it can be like the suggested from @Omikhleia `!format?[alt](path)`
3.  The classes and attributes, that can determine a wide ranges of option like do a histogram from this CSV file.

**Comment by AvtechScientific** [comment]
Any decision/implementation on this topic?

**Comment by jgm** [comment]
I don't think this makes sense.  `![]()` is an inline element; transclusion of blocks of texts is something that would need to be done prior to inline parsing.  If we did add transclusion, I'd favor a different approach.

**Comment by AvtechScientific** [comment]
In the suggested [PR](https://github.com/jgm/djot.js/pull/128/) - we do preprocess before parsing. [preprocessInclusions()](https://github.com/AvtechScientific/djot.js/blob/inclusion/src/include.ts#L277) runs on raw text before parse() is called. The inline parser never sees `![](file.dj)` — it's already been replaced with the file's content. The meaning of `![]()` becomes more general. Instead of only - "dump this image here" - now it's "dump whatever I want here" (images are obviously also covered as backwards compatibility is preserved). Do you prefer a separate syntax, like, e.g. `+[]()`?

**Comment by jgm** [comment]
If it's going to be a preparsing phase, then I think, conceptually, it makes more sense to have a notation that is explicitly marked for processing before parsing. (Compare CPP in C source files.)  And then of course you run into the problem that you won't be able to put these strings in any literal text, even a code block...

**Comment by AvtechScientific** [comment]
Being the developer of djot.js you think in terms of implementation concepts. In that realm image inclusion and text file inclusion are different things and you naturally want this difference to be projected onto the syntax. CPP/C source files example is indeed relevant there.

On the other hand - I try to provide user's perspective. User is not a programmer. User is not bothered by how you achieve your goals technically. User wants a simple to remember, intuitive, coherent and consistent syntax (that's exactly what I loved about djot compared to markdown). That's why I suggest the same syntax for inclusion of images, text files or whatever other file type that we can add in the future.

BTW - to make it even more easier for non tech-savvy users to use djot I contribute some ideas (not implementation) to [djoter](https://gitlab.com/binyamin.pekar/djoter) (hopefully once [Servo](https://servo.org) implements its announced C API it will be possible to make `djoter` cross-platform).

**Comment by jgm** [comment]
Will it be intuitive to users that this might end up being two paragraphs, not one?
```
My text with ![text](included.txt) and some
more stuff.
```
Or that (if the text, interpolated prior to parsing, includes a line ` ``` `) it might convert the entire rest of the document into a code block?

I imagine that users will conceptualize this as an inline item. It's important that the user has the right mental model.

**Comment by bpj** [comment]
Since an alt-text doesn't make much sense for file inclusion maybe `!(included.txt)` or `!<included.txt>` might work as a syntax.

Also note that the included text probably would need to always be parsed as djot markup and inserted as a list of block elements; otherwise a file inclusion in the middle of a list won't make much sense.

**Comment by jgm** [comment]
> Since an alt-text doesn't make much sense for file inclusion

That depends.  If we really wanted a way to include all kinds of content, as this proposal suggests, then it is handy to have alt text for cases where the specified content can't be included in the target format.  (E.g., a movie in a tex document.)

> the included text probably would need to always be parsed as djot markup and inserted as a list of block elements;

I think the proposal here was to have a pre-parsing pass that includes the content.  Parsing the content as block elements would be a different idea. Both have their advantages and disadvantages.  With a preparsing pass, you could even have something like


````
```
some code:
%include myfile.txt
```
````

which would insert the text into a code block.  But of course this would mean that whatever syntax you use for the include (here `%include FILE`) *couldn't be included in a code block* -- an expressive blindspot.

**Comment by AvtechScientific** [comment]
> Will it be intuitive to users that this might end up being two paragraphs, not one?
> 
> ```
> My text with ![text](included.txt) and some
> more stuff.
> ```
>
> Or that (if the text, interpolated prior to parsing, includes a line `` ``` ``) it might convert the entire rest of the document into a code block?
>

Absolutely! By the (new, enhanced) definition all these is possible and expected. And it's not a problem, since the user has control over the included file and he knows what he includes. He can easily remove ` ``` ` from there if it was not his intent.
 
> I imagine that users will conceptualize this as an inline item. It's important that the user has the right mental model.

Only if influenced by their prior experience/knowledge. If you insist to introduce a separate syntax for this - what about `+[alt text]<file_path>` ?

`+` - intuitive for inclusion/addition
`[alt_text]` - coherent with the current image inclusion
`<file_path>` - coherent with `<URI>`

**Comment by AvtechScientific** [comment]
> > Since an alt-text doesn't make much sense for file inclusion
> 
> That depends. If we really wanted a way to include all kinds of content, as this proposal suggests, then it is handy to have alt text for cases where the specified content can't be included in the target format. (E.g., a movie in a tex document.)

Even for included plain text/djot documents alt-text makes sense: just as in www - it will be displayed instead of the file's content if for whatever reason the file could not be read.

> > the included text probably would need to always be parsed as djot markup and inserted as a list of block elements;
> 
> I think the proposal here was to have a pre-parsing pass that includes the content. Parsing the content as block elements would be a different idea. Both have their advantages and disadvantages. With a preparsing pass, you could even have something like
> 
> ```
> ```
> some code:
> %include myfile.txt
> ```
> ```
> 
> which would insert the text into a code block. But of course this would mean that whatever syntax you use for the include (here `%include FILE`) _couldn't be included in a code block_ -- an expressive blindspot.

Yes, in verbatim/code blocks inclusions are interpreted literarily (i.e. without dumping any content). If you do want to include a pre-existing code file without copy/pasting it, we might implement something like this: ` ``` +[my_file]<myfile.txt>`

It probably was not your intent, but I think that using English words in syntax is not a good idea ` %include myfile.txt`. It contradicts the idea of lightweight mark up, as it is "too heavy" and creates problems for documents with non-latin letters.

**Comment by AvtechScientific** [comment]
@jgm - what do you think about introducing `+[alt text]<file_path>` syntax for transclusion (of both plain text/djot files and images) and at the same time deprecating `![alt text](image.png)`. This way - users will not be influenced by their prior experience/expectation that this is an inline element, and on the other hand we will achieve consistency while minimizing syntax that needs to be learned/remembered?

**Comment by jgm** [comment]
> Yes, in verbatim/code blocks inclusions are interpreted literarily (i.e. without dumping any content).

But that means that your "preparsing pass" actually has to do some parsing, at least to identify the verbatim parts.
It's a possible performance issue, because you'll need to parse the document twice.

> It probably was not your intent, but I think that using English words in syntax is not a good idea

Agreed, this was just a placeholder example.

> `+[alt text]<file_path>`

I still have the concern that it looks like inline syntax.

**Comment by AvtechScientific** [comment]
> > Yes, in verbatim/code blocks inclusions are interpreted literarily (i.e. without dumping any content).
> 
> But that means that your "preparsing pass" actually has to do some parsing, at least to identify the verbatim parts. It's a possible performance issue, because you'll need to parse the document twice.

Inclusions inside fenced code blocks are correctly ignored, and it does not require a double pass. Here's how it works:

`splitByCodeFences()` (include.ts, line 84) does a single linear scan splitting input into `{content, isCode}` segments. It tracks ` ``` ` and ` ~~~ ` fence markers with matching close detection.
In `preprocessInclusions()` (line 304), code segments are skipped immediately:

```
if (seg.isCode) {
  accumulatedText += seg.content + "\n";
  return seg.content;  // pass through unchanged
}
```

The INCLUSION_RE regex only runs on non-code segments.
So it's one pass through the text (segment splitting), then inclusion regex only on prose segments. No double parsing.

Note: this only handles fenced code blocks. Inline verbatim spans (backticks) are not filtered, but that's fine — `![](path)` inside backticks would be treated as literal text by the djot parser anyway and wouldn't render as an inclusion.

> > `+[alt text]<file_path>`
> 
> I still have the concern that it looks like inline syntax.

Why?... The syntax is new and you define/assign it a new meaning... Nevermind... Let's assume we have found a new perfect syntax `S` that looks like a block syntax.  But then people will arise and say: "Hey, `S` looks like a block syntax! But when I include an image - I expect it to be an inline syntax!"...

But let's investigate it deeper... I claim that inclusion is neither inline- nor block-element. It's not an element at all. It's a new creature. Intrinsically it's an ACTION rather than an ELEMENT. Action of substitution of an inclusion-instance by an external content (an image, a plaint text file, a .csv file, a movie, whatever). It becomes evident when you look on the generated events (try my branch out with `-t events`) - inclusions do not occupy any space (judging by the position values). Each substitution might end up to be an inline element (like including an image -> `<img>`; or an unknown file type -> `<href>`) but it also might end up to be a block-element (like including an .csv file turning it into a table on the fly; or including other djot/plaintext files). Inclusion event/node indicates that external content must be accessed. So when looking for a syntax for this we should not focus on it looking like an inline or block element, as this is not relevant, but rather make it generic, concise, expressive, intuitive and memorable... Enough with philosophy... Now, @jgm,  let me change the syntax of both image and inclusions to  `+[alt text]<file_path>` and wrap images with inclusion event/node for consistency...

If not - what do you suggest?

**Comment by jgm** [comment]
> But let's investigate it deeper... I claim that inclusion is neither inline- nor block-element. It's not an element at all. It's a new creature. Intrinsically it's an ACTION rather than an ELEMENT

Yes, that is exactly what I was getting at above when I called it a "processing instruction."

> Inclusions inside fenced code blocks are correctly ignored, and it does not require a double pass. Here's how it works:

I can already see a bug.  Fenced code blocks don't always start at the left margin, because they can occur inside list items and other indented constructs.  Keeping track of everything you need to keep track of here is going to require a nontrivial bit of parsing of at least block structure.

And what about inline code?
```
`![file](file.txt)`
```
or, if you are only detecting these at the beginnings of lines (which would be an odd restriction),
```
`blah blah
![file](file.txt)
blah`
```
So it looks like you'd need to correctly parse inline structure too.  And now we're looking at two parsing passes.  Perhaps the first one can be simplified slightly (at the cost of introducing possible bugs -- we don't want things to break, e.g., when we add block comment syntax), but it is still a performance hit.

Also (just for one more example), what about comments in attributes?
```
*hello*{foo="bar"
%
![file](file.txt)
}
```

**Comment by jgm** [comment]
I'd say: people who want a fully general include mechanism, one that can include things in inline or block contents, and even in code blocks, can *already* just use a preprocessor like `cpp` or `m4`.

If we were to add an include syntax to djot, I think it would make sense for it to function differently.  It would be parsed as a block element and would cause the lines of the included file to be included in the stream of lines being parsed.  (And an implementation should also take care to exit with an error if an include "loop" in encountered.)

**Comment by AvtechScientific** [comment]
> > But let's investigate it deeper... I claim that inclusion is neither inline- nor block-element. It's not an element at all. It's a new creature. Intrinsically it's an ACTION rather than an ELEMENT
> 
> Yes, that is exactly what I was getting at above when I called it a "processing instruction."

So are you OK with me  changing the syntax of both image and inclusions to +[alt text]<file_path> and wrap images with inclusion event/node for consistency?...

> 
> > Inclusions inside fenced code blocks are correctly ignored, and it does not require a double pass. Here's how it works:
> 
> I can already see a bug. Fenced code blocks don't always start at the left margin, because they can occur inside list items and other indented constructs. Keeping track of everything you need to keep track of here is going to require a nontrivial bit of parsing of at least block structure.
> 
> ...
>
> ```
> *hello*{foo="bar"
> %
> ![file](file.txt)
> }
> ```

Just added tests to cover all the cases you've mentioned here (all of them within a single pass O(n)).

**Comment by AvtechScientific** [comment]
> I'd say: people who want a fully general include mechanism, one that can include things in inline or block contents, and even in code blocks, can _already_ just use a preprocessor like `cpp` or `m4`.
> 
> If we were to add an include syntax to djot, I think it would make sense for it to function differently. It would be parsed as a block element and would cause the lines of the included file to be included in the stream of lines being parsed. (And an implementation should also take care to exit with an error if an include "loop" in encountered.)

Sure. Preprocessors, for example, can't handle this case:

```
file_a.dj:

# File A
Is *this bold

file_b.dj:

and* this?
# File B

result.dj:

![file_a](file_a.dj) ![file_b](file_b.dj)
```

We have even a special node for this case: InclusionBoundarySpan... And yes - there is a cyclic inclusion detector.

**Comment by jgm** [comment]
I think maybe we should talk at the level of goals, instead of getting bogged down in the details of one particular proposal.  **Why is an include syntax desired, and what features does it need to have?**

1. Should the include syntax work even inside contexts like code blocks, code spans, attributes, comments?  If so, this points in the direction of some kind of preprocessor.  If not:

2. Should it be possible to include text in an inline content (as your example above is meant to illustrate), or only at the block level?  What are the real-world needs for inline-level includes?  Would they be better handled with some kind of templating library?

3. Should includes be a built-in feature of the syntax and djot parsers, or can we handle them with a filter that replaces a certain AST element (e.g. a specially marked code block or div)?  This approach would work if the included content can be parsed on its own, independently of the parent document, but would not work if e.g. the parent document relies on link definitions defined in the included file.
 
4. What should the syntax be (to be discussed last).

I don't think it's obvious that we *need* an include syntax (I've written books and many websites with pandoc's markdown and not missed it; I haven't even felt the need to reach for the simple include filters that are available). We should keep in mind that simplicity is an important goal for djot. The [patch in djot.js#128](https://github.com/jgm/djot.js/pull/128/changes#diff-a2934fe8b80398a33b3b4e41deafebe030d525d108b78b0a8568b84aadc140ea) is quite complex and has just needed to be modified to deal with several cases I pointed out without even looking much at the code.  Do we want to force every djot parser to implement such logic?

Perhaps, though, the general discussion should move to the existing, more general issue https://github.com/jgm/djot/issues/199
